src/main/java/Domain/ShoppingBasket.java

```diff
@@ -33,6 +36,33 @@ public double getShoppingBasketPrice() {
         return _basketTotalAmount;
     }
 
+    /*
+     * Go through the list of products in the basket and purchase them.
+     * If an exception is thrown, cancel the purchase of all the products that were bought.
+     * This function only updates the item's stock.
+     */
+    public void purchaseBasket() {
+        List<Product> boughtProductList = new ArrayList<>();
+        for (Product product : _productList) {
+            try{
+                product.purchaseProduct();
+                boughtProductList.add(product);
+            }
+            catch (ItemOutOfStockExepction e){
+                for (Product boughtProduct : boughtProductList) {
+                    boughtProduct.cancelPurchase();
+                }
+                throw new ItemOutOfStockExepction("One of the products in the basket is out of stock");
```

Comment:
why exception and not worning? do we want the system to stop? or just inform the user on the items he has? if on items left at all in the stock of maybe error

Reply:
I dont understand

Reply:
I can change it to return boolean and make it so it wont thorw exception but I am handeling the exception in the function that calls this function, I change the signatures to throw exception for now so we wont forget it.

Reply:
I agree its not the best idea but I think its the best for now, I will change it in future if needed.

Reply:
Done.